### PR TITLE
🌱 Add test utils for cluster class

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This is part of the activities for the implementation of the [ClusterClass proposal](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/202105256-cluster-class-and-managed-topologies.md). 

This PR add a first set of test utils/utils that will simplify the work for https://github.com/kubernetes-sigs/cluster-api/issues/4970 and https://github.com/kubernetes-sigs/cluster-api/issues/4971 (and also subsequent PRs)

@killianmuldoon

### CODE REVIEW

File: controllers/clustertopology_controller_compute.go
Comment (stmcginnis) [inline controllers/clustertopology_controller_compute.go:26]: Minor nit that I think only needs to be addressed if you need to do an update for any other reason:

s/and it that case/and in that case/

File: controllers/clustertopology_controller_compute.go
Comment (sbueringer) [reply controllers/clustertopology_controller_compute.go:37]: Or maybe I misinterpreted the intention (i.e. the godoc) and we always validate that we have a full object ref before calling this func?

File: controllers/clustertopology_controller_compute_test.go
Comment (sbueringer) [inline controllers/clustertopology_controller_compute_test.go:45]:
```suggestion
	t.Run("Gets nil if the nested ref does not exist", func(t *testing.T) {
```
nit

**Comment by sbueringer** [comment]
@fabriziopandini Thank you, looks great! I'll start using it tomorrow on my PR.

File: controllers/clustertopology_controller_compute.go
Comment (fabriziopandini) [reply controllers/clustertopology_controller_compute.go:37]: Agree we should make this more robust.

File: controllers/clustertopology_controller_compute.go
Comment (sbueringer) [inline controllers/clustertopology_controller_compute.go:28]: In case one of the fields is not set, we could either return:
* a partial ObjectReference (current impl)
* nil

I wonder what's easier to handle for the consumers. A concrete example:

I try to retrieve the optional InfrastructureMachineTemplate from KubeadmControlPlane via getNestedRef

The following could happen:
* ObjectRef not set => `getNestedRef` retuns nil => I assume the ref doesn't exist on the ControlPlane
* ObjectRef is set => `getNestedRef` returns an ObjectReference => I handle the MachineTemplate
* ObjectRef is partially set => Now we have two options for the implementation:
    * => `getNestedRef` returns nil => I assume the ref doesn't exist on the ControlPlane
    * => `getNestedRef` returns a partial ObjectReference => I have to either just use the ObjectReference and the Get will fail or I have to validate if I got a "full" ObjectReference and handle it accordingly

So overall: I just wanted to ask if we just want to interpret a partial ObjectReference the same way as a non-existing one (consistently), then we wouldn't have to do anything on the consumer side when we return nil instead of a partial ObjectReference.

Comment (fabriziopandini) [reply controllers/clustertopology_controller_compute.go:28]: FYI we are enforcing reference should be valid/complete in https://github.com/kubernetes-sigs/cluster-api/issues/4940

Assuming the above context, this func should get what it founds and we can reasonably assume the value is correct and complete; If we are going to find out this assumption too week while implementing consumers in the upcoming PRs, we can iterate and improve accordingly

Comment (sbueringer) [reply controllers/clustertopology_controller_compute.go:28]: I think #4940 does not include the ref in ControlPlane. But I think that's the responsibility of the control plane provider and otherwise they will just get an error here which is the best we can do in the controller code anyway.

So agree, let's keep it as is.

Related follow up question: Maybe I'm missing something but could it be that KCP only validates 3/4 fields of the ObjectReference?
https://github.com/kubernetes-sigs/cluster-api/blob/master/controlplane/kubeadm/api/v1alpha4/kubeadm_control_plane_webhook.go#L280-L309
(our 4 relevant fields are all optional in the OpenAPI schema)

In case I'm right, I can open a PR to add the validation for `Name`

File: controllers/clustertopology_controller_compute.go
Comment (stmcginnis) [inline controllers/clustertopology_controller_compute.go:34]: We don't do anything with err if err != nil. Would it be worth adding a comment here explaining why we drop that?

Comment (fabriziopandini) [reply controllers/clustertopology_controller_compute.go:34]: It is already described in the func description that this field is optional (and in that case we return nil)

**Comment by stmcginnis** [comment]
/lgtm

Though looks like this needs to be rebased now.

File: controllers/clustertopology_controller_compute.go
Comment (fabriziopandini) [reply controllers/clustertopology_controller_compute.go:28]: Validating name was should go in #4940; please comment there

**Comment by fabriziopandini** [comment]
/holde
We are changing a little bit the APIs (@vincepri is opening a PR ATM), and this will make explicit when a controlPlane template has a reference so we can drop getNestedRef or to scope it usages only to tests

**Comment by fabriziopandini** [comment]
/hold cancel

Given  #5000 `getNestedRef` should not be used anymore for inspection so I enforced checks that the reference should be “complete”

@sbueringer @killianmuldoon ^^

File: controllers/clustertopology_controller_compute_test.go
Comment (sbueringer) [inline controllers/clustertopology_controller_compute_test.go:46]:
```suggestion
	t.Run("Get fails if the nested ref does not exist", func(t *testing.T) {
```
nit

File: controllers/clustertopology_controller_compute_test.go
Comment (sbueringer) [inline controllers/clustertopology_controller_compute_test.go:45]: Maybe add one other test case where the objectref partially exists

**Comment by sbueringer** [comment]
@fabriziopandini @killianmuldoon I'm extending the fake objects currently on my PR to cover the whole functionality.

File: controllers/clustertopology_controller_compute_test.go
Comment (ykakarap) [inline controllers/clustertopology_controller_compute_test.go:38]: Since this is a contrived example within a test case, it probably doesnt matter as much but for an actual controlplanetemplate the path to the infrstructureRef will actually be `spec.template.spec.machineTemplate.infrastructureRef`, isnt it?

Comment (sbueringer) [reply controllers/clustertopology_controller_compute_test.go:38]: Yes.

**Review by vincepri — APPROVED** [review]
/approve
/lgtm

**Comment by sbueringer** [comment]
@fabriziopandini  I think this has to be rebased after the merge of #5000 and fixed. Iirc just a minor fix in `fakeClusterClass`

**Comment by fabriziopandini** [comment]
@sbueringer @vincepri rebased on top of #5000 (control plane machine infrastructure is now an explicit field in ClusterClass)

**Pull request merged**
